### PR TITLE
Lower PtrToInst instructions.

### PR DESCRIPTION
**Pull request by ptersilie**

### CODE REVIEW

File: llvm/lib/YkIR/YkIRWriter.cpp [hunk 1]
```diff
@@ -889,6 +889,23 @@ class YkIRWriter {
     InstIdx++;
   }
 
+  /// Serialise ptrtoint instruction.
+  void serialisePtrToIntInst(PtrToIntInst *I, FuncLowerCtxt &FLCtxt, unsigned BBIdx,
+                         unsigned &InstIdx) {
+    // opcode:
+    serialiseOpcode(OpCodeCast);
+    // cast_kind:
+    serialiseCastKind(CastKindZeroExt);
```
Comment (vext01) [inline llvm/lib/YkIR/YkIRWriter.cpp:898]: Just reading the LLVM reference, `ptrtoint` has more semantics than we first thought:
https://llvm.org/docs/LangRef.html#ptrtoint-to-instruction

 - We should reject the vector variant (encode an unimplemented instruction). You can copy/adapt existing code from other opcodes that have vector variants.
 - `ptrtoint` can truncate to a smaller integer size, so zero extension isn't generally correct. If you don't want to handle those now, you can make an unimplemented instruction if the output type isn't pointer sized.

Comment (vext01) [reply llvm/lib/YkIR/YkIRWriter.cpp:898]: (when rejecting certain versions of instructions, I've also been adding a case to `tests/ir_lowering/unsupported_variants.ll` to check they are properly rejected)

Comment (ptersilie) [reply llvm/lib/YkIR/YkIRWriter.cpp:898]: My first instinct after your comment was to just add a `aot_ir::PtrToIntInst`. But reading the langref, it does say they either truncate/zeroextend/noop anyway. So we can probably do without such an instruction and just lower each case to different existing aot instructions.

Comment (vext01) [reply llvm/lib/YkIR/YkIRWriter.cpp:898]: that would be fine by me.

**Comment by ptersilie** [comment]
I've updated the PR.

**Comment by ptersilie** [comment]
Here you go.

**Comment by vext01** [comment]
Perfect. Please squash.

**Comment by ptersilie** [comment]
Done.

**Comment by ptersilie** [comment]
Ok to squash?

**Comment by vext01** [comment]
Please squash

**Comment by ptersilie** [comment]
Done.

**Pull request merged**